generator/findby.go

```diff
@@ -0,0 +1,192 @@
+		"    return q.Where(cond(Schema.%[4]s.%[1]s, v))\n" +
+		"}"
+	tplFindByID = "\n\n// FindBy adds a new filter to the query that will require that\n" +
+		"// the %[1]s property is equal to one of the passed values; if no passed values, it will do nothing\n" +
+		"func (q *%[2]s) FindBy%[1]s(v ...%[3]s) *%[2]s {\n" +
+		"    if len(v) == 0 {return q}\n" +
+		"    values := make([]interface{}, len(v))\n" +
+		"    for i, val := range v {values[i] = val}\n" +
+		"    return q.Where(kallax.In(Schema.%[4]s.%[1]s, values...))\n" +
+		"}"
+)
+
+// isValid returns true if the passed type is valid for an autogenerated FindBy
+type isValid func(pkg *types.Package, typ types.Type) bool
+
+var typeValidators = []isValid{isBasicType, isSpecialType}
+
+// GenFindBy generates FindByPropertyName for all model properties that are
+// valid types or collection of valid types.
+func (td *TemplateData) GenFindBy(model *Model) string {
+	var buf bytes.Buffer
+	td.genFindBy(&buf, model, model.Fields)
+	return buf.String()
+}
+
+func (td *TemplateData) genFindBy(buf *bytes.Buffer, parent *Model, fields []*Field) {
+	for _, f := range fields {
+		switch {
+		case f.Inline():
+			td.genFindBy(buf, parent, f.Fields)
+		case f.Name == "ID":
+			writeTplCode(buf, parent, f, tplFindByID)
+		case isEqualizable(f):
+			writeTplCode(buf, parent, f, tplFindByEquality)
+		case isSortable(f):
+			writeTplCode(buf, parent, f, tplFindByCondition)
+		case isCollection(f):
+			writeTplCode(buf, parent, f, tplFindByInclude)
+		}
+	}
+}
+
+func writeTplCode(buf *bytes.Buffer, parent *Model, f *Field, tpl string) {
+	validType, err := validType(f)
+	if err != nil {
+		return
+	}
+
+	property := f.Name
+	query := parent.QueryName
+	model := parent.Name
+	buf.WriteString(fmt.Sprintf(tpl, property, query, validType, model))
+}
+
+// validType returns the short name of the type that can be used to search for
+// the passed Field, in an autogenerated 'FindBy';
+// if the Field is not valid for an autogenerated 'FindBy' it will return an error
+func validType(f *Field) (string, error) {
+	analycedType := f.Node.Type()
+	collectionAlreadyScanned := false
+	for {
+		valid, deepest := lookupValidRecursively(f.Node.Pkg(), analycedType)
+		if valid != nil {
+			return getShortName(analycedType), nil
+		}
+
+		if collectionAlreadyScanned {
+			break
+		}
+
+		singular := singularType(deepest)
+		if singular == nil {
+			break
+		}
+
+		analycedType = singular
+		collectionAlreadyScanned = true
+	}
+
+	return "", fmt.Errorf("kallax: '%s' is not valid for FindBy", f.Node.Type())
+}
+
+// lookupValid returns the first valid type looking into the underlying types of
+// the passed one; if no valid found it will return the deepest underlying
+func lookupValidRecursively(pkg *types.Package, typ types.Type) (valid, deepest types.Type) {
+	for _, isValid := range typeValidators {
+		if isValid(pkg, typ) {
+			return typ, nil
+		}
+	}
+
+	underlaying := typ.Underlying()
```

Comment:
#epicFail, I just used the right word at the right
`underlaying := typ.Underlying()`